### ISSUE TITLE
[WIP]Checkpoint state_dict and enable resharding for Sharded Tensor

### ISSUE THREAD

**Issue by zzzwen** [pytorch/pytorch#76386]
### 🚀 The feature, motivation and pitch

with @kumpera  @pritamdamania  @aazzolini  @xunnanxu @yifuwang  

## TL;DR
The proposal here is to introduce a default solution to save, load and resharding ShardedTensor. It will complete the ecosystem, and offer users the "battery included" optional when working with  ShardedTensor. 

### For the WIP code, see this PR https://github.com/pytorch/pytorch/pull/76123

* ### storage_writer / storage_reader

StorageWriter / StorageReader 

To provide a solution for resharding ShardTensor, we need to define interface classes that define the storage contract, 
How StorageWriter/Reader actually write to or read from the underlying storage is out of the scope of `torch.distributed`, the subclass can optimize per its use case. 
We will implement a default subclass to support file system. And there can be subclass that optimized for AWS, GCP, Azure and etc.  
It is also possible to subclass them for features, like a QuantizedStoragesWriter can be implemented to write quantized checkpoint.

```
class StorageWriter():
   def write_metadata(self, metadata: ModelMetadata):
       """
       torch.distributed generates the ModelMetadata.
       The StorageWriter can extend it with storage specific metadata.
       And its "Reader" counterpart will need to strip the extra metadata
       and returns the original "ModelMetadata".
       """
       ...

    def write_bytes(self, reqs: List[BytesWriteRequest]) -> torch.Future:
       """
       torch.distributed will issue all the requests at once.
       StorageWriter has the option to batch/cache.
       torch.distributed will wait on the Future it returns.
       and this function should return WITHOUT waiting on all write to storage to finish
       """
       ...

    def write_tensors(self, reqs: List[TensorWriteRequest]) -> torch.Future:
       """
       torch.distributed will issue all the requests at once.
       StorageWriter has the option to batch/cache.
       torch.distributed will wait on the Future it returns.
       and this function should return WITHOUT waiting on all write to storage to finish
       """
       ...
```

```
class StorageReader():
   def read_metadata(self) -> ModelMetadata:
       """
       It needs to match  add additional metadata
       Providing the it counterpart "StorageReader" will strip the additional
       metadata and returns the original "ModelMetadata"
       """
       ...

   def read_bytes(self, req: List[BytesReadRequest]) -> torch.Future:
       """
       torch.distributed will issue all the requests at once.
       StorageReader has the option to batch/cache.
       torch.distributed will wait on all the Futures insides.
       and this function should return WITHOUT waiting on all read from storage to finish
       """
       ...

   def read_tensors(self, req: List[TensorReadRequest]) -> torch.Future:
       """
       torch.distributed will issue all the requests at once.
       StorageReader has the option to batch/cache.
       torch.distributed will wait on all the Futures insides.
       and this function should return WITHOUT waiting on all read from storage to finish
       """
       ...
```

* ### ReadRequest/WriteRequest.

The storage layer will operate on ReadRequest/WriteRequest.

There are two types of requests: 
1. For arbitrary objects.
The request will contain bytes. It is essential for primitive types, like the `step` in optimizer state, which is usually an int.
Note that we CANNOT support resharding for arbitrary bytes. 

2. For torch.Tensor or torch.distributed.ShardedTensor 
The requests will contain tensor(view). The storage writer or reader is expected to handle the data correctly, which includes moving the storage from/to the correct device. 

```
@dataclass
class BytesWriteRequest:
  buffer: BytesIO
  storage_key: str

@dataclass
class TensorWriteRequest:
  tensor: torch.Tensor
  storage_key: str

@dataclass
class BytesReadRequest:
  buffer: BytesIO
  storage_key: str

@dataclass
class TensorReadRequest:
  tensor: torch.Tensor   
  storage_key: str
  offsets: Tuple[int, ...]
  lengths: Tuple[int, ...]
```

* ### `torch.distributed.save_state_dict/load_state_dict`

```
def save_state_dict( state_dict: Dict[str, Any], storage_writer: StorageWriter) -> None:
    ...

def load_state_dict( state_dict: Dict[str, Any],  storage_reader: StorageReader) -> None:
    ...
```
 
`torch.distributed.save_state_dict/load_state_dict` is a layer between the trainer integration.

The state_dict can come from `nn.module.state_dict()`,  it can be a copy of that, or simply something the caller created. It does not assume anything.

The state_dict's key is the FQN we usually get from calling nn.Module.name_parameters() e.g. `model.sub_module.linear.bias`. Its value can be Tensor, ShardedTensor or Any. 

Note that during read, it will write back to the input `state_dict` *inplace*.  
For ShardedTensors, it will attempt to reshard iff needed.
For Tensor, it will load into the tensor storage iff the size match.
For Any, it will replace the value in the state_dict, without any validation.

## Sample Code for different usage 

- ### Distributed checkpointing

```
# Assume SPMD 
my_model = MyModule()
# We must call this function prior to state_dict()
my_model._register_state_dict_hook(state_dict_hook)
model_state_dict = my_model.state_dict()

if dist.get_rank() == 0:
    fs_storage_writer = torch.distributed._checkpoint.FileSystemWriter("/checkpoint/1")
else:
    fs_storage_writer= None

fs_storage_writers = [fs_storage_writer]
dist.broadcast_object_list(fs_storage_writers)

torch.distributed._checkpoint.save_state_dict(
    state_dict=model_state_dict,
    storage_writer=fs_stroage_writer,
)
```

- ### Async Checkpoint 

Assume we have an `InMemoryCache` class that can save a copy of the state_dict in memory
```
# Assume SPMD 
my_model = MyModule()

# We must call this function prior to state_dict()
my_model._register_state_dict_hook(state_dict_hook)
model_state_dict = my_model.state_dict()

if dist.get_rank() == 0:
    fs_storage_writer = torch.distributed._checkpoint.FileSystemWriter("/checkpoint/1")
else:
    fs_storage_writer= None

fs_storage_writers = [fs_storage_writer]
dist.broadcast_object_list(fs_storage_writers)

# The in memory cache is not shared across ranks
in_memory_cache = torch.distributed._checkpoint.InMemoryCache()

# Offload the tensors from the nn.module to in memory cache, this should be fast
torch.distributed._checkpoint.save_state_dict(
    state_dict=model_state_dict,
    storage_writer=in_memory_cache,
)

# Run the write to file system in another process.
p = Process(
    target=torch.distributed._checkpoint.save_state_dict, 
    args=(in_memory_cache.state_dict, fs_storage_writers)
)
p.start()
...
# Continue training
```

     

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by kumpera** [comment]
This feature has being merged.
